src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+            raise TypeError("The loss function must a Callable!")
+
+        if (not hasattr(func, "__module__")) or (
+            not (
+                func.__module__.startswith("jax")
+                or func.__module__.startswith("neurostatslib.glm")
+            )
+        ):
+            raise ValueError(
+                f"The function {func.__name__} is not from the jax namespace. "
+                "Only functions from the jax namespace are allowed."
+            )
+
+    @abc.abstractmethod
+    def instantiate_solver(
+        self,
+        loss: Callable[
+            [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jnp.ndarray
+        ],
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """Abstract method to instantiate a solver with a given loss function."""
+        pass
+
+    def get_runner(
+        self,
+        solver_kwargs: dict,
+        run_kwargs: dict,
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """
+        Get the solver runner with provided arguments.
+
+        Parameters
+        ----------
+        solver_kwargs :
+            Additional keyword arguments for the solver instantiation.
+        run_kwargs :
+            Additional keyword arguments for the solver run.
+
+        Returns
+        -------
+        :
+            The solver runner.
+        """
+        solver = getattr(jaxopt, self.solver_name)(**solver_kwargs)
+
+        def solver_run(
+            init_params: Tuple[jnp.ndarray, jnp.ndarray], X: jnp.ndarray, y: jnp.ndarray
+        ) -> jaxopt.OptStep:
+            return solver.run(init_params, X=X, y=y, **run_kwargs)
+
+        return solver_run
+
+
+class UnRegularizedSolver(Solver):
+    """
+    Solver class for optimizing unregularized models.
+
+    This class provides an interface to various optimization methods for models that
+    do not involve regularization. The optimization methods that are allowed for this
+    class are defined in the `allowed_optimizers` attribute.
+
+    Attributes
+    ----------
+    allowed_optimizers : list of str
+        List of optimizer names that are allowed for this solver class.
+
+    Methods
+    -------
+    instantiate_solver(loss)
+        Instantiates the optimization algorithm with the given loss function.
+
+    See Also
+    --------
+    [Solver](./#neurostatslib.solver.Solver) : Base solver class from which this class inherits.
+    """
+
+    allowed_optimizers = [
+        "GradientDescent",
+        "BFGS",
+        "LBFGS",
+        "ScipyMinimize",
+        "NonlinearCG",
+        "ScipyBoundedMinimize",
+        "LBFGSB",
+    ]
+
+    def __init__(
+        self, solver_name: str = "GradientDescent", solver_kwargs: Optional[dict] = None
+    ):
+        super().__init__(solver_name, solver_kwargs=solver_kwargs)
+
+    def instantiate_solver(
```

Comment:
I feel like `instantiate_solver()` can be moved to the superclass, with the four lines found here, and then the regularized ones can over-write, and just pass up the penalized loss instead of their regular loss. Thoughts?

Reply:
I changed it to a concrete method, I ended up needing to overriding it for ridge and proximal solvers. Let me know if we can further simplify the logic